tutorials/03_plugins.md

```diff
@@ -3,121 +3,218 @@
+/// \brief Destructor
+public: virtual ~HelloPlugin();
+```
 
-### Topic echo
+### Implementation of the constructor and the destructor.
 
-Echo messages from a Gazebo Transport topic.
+The constructor simply calls the base class constructor:
 
-    gz gui -c examples/config/pubsub.config
+```cpp
+HelloPlugin::HelloPlugin()
+  : gz::gui::Plugin()
+{
+}
+```
 
-### Key publisher
+The destructor is empty because no cleanup is required as such.
 
-Publish key presses through Gazebo Transport.
+```cpp
+HelloPlugin::~HelloPlugin()
+{
+}
+```
 
-    gz gui -c examples/config/keypublisher.config
+### The Configuration File
 
-Change the topic to echo to `/keyboard/keypress`, echo, and start pressing keys.
+A plugin may require certain configuration files to specify certain parameter values.
+
+In this case, we would want the configuration file to hold the message to be printed on the display (here the display is the terminal) when the `Hello, Plugin!` button is clicked.
+
+Thus, the [HelloPlugin.config](https://github.com/gazebosim/gz-gui/blob/main/examples/plugin/hello_plugin/HelloPlugin.config) looks like this:
+
+```xml
+<plugin filename="HelloPlugin">
+  <message>Hello, plugin config!</message>
+</plugin>
+```
+
+The `LoadConfig` function tries to get the message from the message tag.
+
+```cpp
+void HelloPlugin::LoadConfig(const tinyxml2::XMLElement *_pluginElem)
+{
+  if (!_pluginElem)
+    return;
+
+  // Take parameters from XML at runtime
+  auto messageElem = _pluginElem->FirstChildElement("message");
+  if (nullptr != messageElem && nullptr != messageElem->GetText())
+    this->message = messageElem->GetText();
+}
+```
+
+We have the `OnButton()` function which would print the message when clicked.
+
+```cpp
+void HelloPlugin::OnButton()
+{
+  std::cout << this->message << std::endl;
+}
+```
+
+Thereafter, `GZ_ADD_PLUGIN` registers the plugin with the Gazebo GUI system, making it discoverable at runtime.
+
+Refer to the [documentation](https://gazebosim.org/api/plugin/2/gz_2plugin_2Register_8hh.html).
+
+```cpp
+GZ_ADD_PLUGIN(HelloPlugin,
+              gz::gui::Plugin);
+```
+
+## Configuring plugins
+
+As shown above, Gazebo GUI supports loading XML configuration files, which may contain
+parameters to be passed directly to the plugin. See the \ref config
+tutorial for more information.
+
+## Finding plugins
+
+Gazebo GUI will look for plugins on the following paths, in this order:
+
+1. All paths set on the `GZ_GUI_PLUGIN_PATH` environment variable
+2. All paths added by calling `gz::gui::App()->AddPluginPath("<<<custom_path>>>");`. Here `App()` is the pointer to running application in the `gz::gui` namespace.
+3. `~/.gz/gui/plugins`
+4. [Plugins which are installed with Gazebo GUI](https://github.com/gazebosim/gz-gui/tree/gz-gui9/src/plugins)
+
+## Setting up the environment for custom plugins
+
+If you have built `gz-gui` from source, follow the steps below:
+
+1. Navigate to the directory where gz-gui is present.
```

Comment:
Users should be able to build and run their custom plugins without having to build gz-gui from source. The `GZ_GUI_PLUGIN_PATH` environment variable just needs to point to the directory containing their custom plugin, e.g. for the Hello plugin example, it'll just be in its `build` directory. So suggestion below to rephrase the instructions:

```suggestion
To run your custom plugin, follow the steps below:

1. Navigate to the directory where your custom plugin is built.
```